engine/render/src/render/buffer.cpp

```diff
@@ -0,0 +1,146 @@
+// https://www.defold.com/license
+//
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#include "render_private.h"
+
+namespace dmRender
+{
+    static HRenderBuffer NewRenderBuffer(dmGraphics::HContext graphics_context, RenderBufferType type)
+    {
+        switch(type)
+        {
+            case RENDER_BUFFER_TYPE_VERTEX_BUFFER: return (HRenderBuffer) dmGraphics::NewVertexBuffer(graphics_context, 0, 0x0, dmGraphics::BUFFER_USAGE_DYNAMIC_DRAW);
+            case RENDER_BUFFER_TYPE_INDEX_BUFFER:  return (HRenderBuffer) dmGraphics::NewIndexBuffer(graphics_context, 0, 0x0, dmGraphics::BUFFER_USAGE_DYNAMIC_DRAW);
+            default:assert(0);
+        }
+        return (HRenderBuffer) 0;
+    }
+
+    static void DeleteRenderBuffer(HRenderBuffer buffer_to_remove, RenderBufferType type)
+    {
+        switch(type)
+        {
+            case RENDER_BUFFER_TYPE_VERTEX_BUFFER:
+                dmGraphics::DeleteVertexBuffer((dmGraphics::HVertexBuffer) buffer_to_remove);
+                break;
+            case RENDER_BUFFER_TYPE_INDEX_BUFFER:
+                dmGraphics::DeleteIndexBuffer((dmGraphics::HIndexBuffer) buffer_to_remove);
+                break;
+            default:assert(0);
+        }
+    }
+
+    static inline void CreateAndPush(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        buffer->m_Buffers.Push(NewRenderBuffer(render_context->m_GraphicsContext, buffer->m_Type));
+    }
+
+    static inline void PopAndDelete(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        DeleteRenderBuffer(buffer->m_Buffers.Back(), buffer->m_Type);
+        buffer->m_Buffers.Pop();
+    }
+
+    HBufferedRenderBuffer NewBufferedRenderBuffer(HRenderContext render_context, RenderBufferType type)
+    {
+        BufferedRenderBuffer* buffer = new BufferedRenderBuffer();
+        memset(buffer, 0, sizeof(BufferedRenderBuffer));
+        RewindBuffer(render_context, buffer);
+        buffer->m_Type = type;
+        return buffer;
+    }
+
+    void DeleteBufferedRenderBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        if (!buffer)
+            return;
+
+        for (int i = 0; i < buffer->m_Buffers.Size(); ++i)
+        {
+            DeleteRenderBuffer(buffer->m_Buffers[i], buffer->m_Type);
+        }
+        delete buffer;
+    }
+
+    HRenderBuffer AdvanceRenderBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        if (!buffer)
+            return 0;
+
+        if (buffer->m_Buffers.Empty())
+        {
+            buffer->m_Buffers.OffsetCapacity(1);
+            CreateAndPush(render_context, buffer);
+        }
```

Comment:
If NewRenderBuffer() already added a buffer, we wouldn't have to do this extra step here, which imo makes it harder to read.

Perhaps shift it to make it only only depend on the possibility to grow:
```
HRenderBuffer AddRenderBuffer(...)
{
    if (render_context->m_MultiBufferingRequired)
    {
        if (buffer->m_Buffers.Full())
        {
            buffer->m_Buffers.OffsetCapacity(4); // The pointers are small in comparison, the allocation itself is likely more costly
        }
        CreateAndPush(render_context, buffer);
        buffer->m_BufferIndex++;
    }
    return buffer->m_Buffers[buffer->m_BufferIndex];
}
```